core/src/consensus.rs

```diff
@@ -345,37 +358,66 @@ where
 
 	// adjust time delta toward goal subject to dampening and clamping
 	let adj_ts = clamp(
-		damp(ts_delta, BLOCK_TIME_WINDOW, DIFFICULTY_DAMP_FACTOR),
+		damp(ts_delta, BLOCK_TIME_WINDOW, DMA_DAMP_FACTOR),
 		BLOCK_TIME_WINDOW,
 		CLAMP_FACTOR,
 	);
 	// minimum difficulty avoids getting stuck due to dampening
-	let difficulty = max(MIN_DIFFICULTY, diff_sum * BLOCK_TIME_SEC / adj_ts);
+	let difficulty = max(MIN_DMA_DIFFICULTY, diff_sum * BLOCK_TIME_SEC / adj_ts);
 
 	HeaderInfo::from_diff_scaling(Difficulty::from_num(difficulty), sec_pow_scaling)
 }
 
+/// Difficulty calculation based on a Weighted Target Exponential Moving Average
+/// of difficulty, using the ratio of the last block time over the target block time.
+pub fn next_wtema_difficulty<T>(_height: u64, cursor: T) -> HeaderInfo
+where
+	T: IntoIterator<Item = HeaderInfo>,
+{
+	let mut last_headers = cursor.into_iter();
+
+	// last two headers
+	let last_header = last_headers.next().unwrap();
+	let prev_header = last_headers.next().unwrap();
+
+	let last_block_time: u64 = last_header.timestamp - prev_header.timestamp;
+
+	let last_diff = last_header.difficulty.to_num();
+
+	// wtema difficulty update
+	let next_diff =
+		last_diff * WTEMA_HALF_LIFE / (WTEMA_HALF_LIFE - BLOCK_TIME_SEC + last_block_time);
```

Comment:
Took me a bit to process so putting a note here: in terms of the RFC, `WTEMA_HALF_LIFE` corresponds to `W * T`. So in this PR we have a `W` of 240, meaning the difficulty can change by a factor `~e` every 240 blocks (4 hours, hence the name of `WTEMA_HALF_LIFE`) 👍

Reply:
correct. the name HALF_LIFE is a bit misleading in that sense, since it corresponds not to a doubling.